server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -185,15 +187,26 @@ public void setAggregatedDfs(AggregatedDfs aggregatedDfs) {
 
     @Override
     public Query rewrite(Query original) throws IOException {
+        Timer concurrentPathRewriteTimer = null;
         if (profiler != null) {
-            profiler.startRewriteTime();
+            if (fromConcurrentPath) {
+                concurrentPathRewriteTimer = new Timer();
+                profiler.getConcurrentPathRewriteTimers().add(concurrentPathRewriteTimer);
+                concurrentPathRewriteTimer.start();
+            } else {
+                profiler.startRewriteTime();
```

Comment:
So `QueryProfiler` already knows if it is part of the concurrent search or not, the logic should be part of the `profiler.startRewriteTime()` and not `ContextIndexSearcher` I believe

Reply:
Resonating with @reta here. I think we can create a separate `ConcurrentQueryProfiler` for concurrent search case and it will override the needed functions in `QueryProfiler` class like for `startRewriteTime`, `getQueryBreakdown` to handle the concurrent search usecase. That profiler can choose to also have a `ThreadAwareQueryProfileTree` which can keep track of threadId to `ConcurrentQueryProfileTree` mapping and also the threadId to rewriteTimers. `ThreadAwareQueryProfileTree` will override the methods of `AbstractQueryProfileTree` and make use of its internal map to return existing or new `ConcurrentQueryProfileTree` instance. Also this way the flag `forConcurrentPath` which is actually differentiating if the query is processing leaves or not (or actual search path) will probably be not needed.

Reply:
Thank you @reta and @sohami for taking the time to review. I agree that this will make the code appear cleaner, and the change is now ready for another review.